batch/sql/add-user-resources-ready-cores-token.sql

```diff
@@ -0,0 +1,176 @@
+    SUM(running_cores_mcpu) as running_cores_mcpu
+  FROM tmp_resources
+  WHERE closed
+  GROUP BY user, token) AS t
+ON user_resources.user = t.user AND user_resources.token = t.token
+SET
+  user_resources.n_ready_jobs = t.n_ready_jobs,
+  user_resources.n_running_jobs = t.n_running_jobs,
+  user_resources.ready_cores_mcpu = t.ready_cores_mcpu,
+  user_resources.running_cores_mcpu = t.running_cores_mcpu;
+
+INSERT INTO batches_staging (batch_id, token, n_jobs, n_ready_jobs, ready_cores_mcpu)
+  SELECT tmp_resources.batch_id,
+    tmp_resources.token,
+    tmp_resources.n_jobs,
+    tmp_resources.n_ready_jobs,
+    tmp_resources.ready_cores_mcpu
+  FROM tmp_resources
+  WHERE NOT closed;
+
+DELIMITER $$
+
+DROP TRIGGER IF EXISTS jobs_after_update;
+CREATE TRIGGER jobs_after_update AFTER UPDATE ON jobs
+FOR EACH ROW
+BEGIN
+  DECLARE in_user VARCHAR(100);
+  DECLARE rand_token INT;
+
+  SELECT user INTO in_user from batches
+  WHERE id = NEW.batch_id;
+
+  SET rand_token = FLOOR(RAND() * @n_tokens);
+
+  IF OLD.state = 'Ready' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_ready_jobs = n_ready_jobs - 1,
+      ready_cores_mcpu = ready_cores_mcpu - OLD.cores_mcpu;
+
+    INSERT INTO ready_cores (token, ready_cores_mcpu) VALUES (rand_token, 0)
+    ON DUPLICATE KEY UPDATE
+      ready_cores_mcpu = ready_cores_mcpu - OLD.cores_mcpu;
+  END IF;
+
+  IF NEW.state = 'Ready' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_ready_jobs = n_ready_jobs + 1,
+      ready_cores_mcpu = ready_cores_mcpu + NEW.cores_mcpu;
+
+    INSERT INTO ready_cores (token, ready_cores_mcpu) VALUES (rand_token, 0)
+    ON DUPLICATE KEY UPDATE
+      ready_cores_mcpu = ready_cores_mcpu + NEW.cores_mcpu;
+  END IF;
+
+  IF OLD.state = 'Running' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_running_jobs = n_running_jobs - 1,
+      running_cores_mcpu = running_cores_mcpu - OLD.cores_mcpu;
+  END IF;
+
+  IF NEW.state = 'Running' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_running_jobs = n_running_jobs + 1,
+      running_cores_mcpu = running_cores_mcpu + NEW.cores_mcpu;
+  END IF;
+END $$
+
+DROP PROCEDURE IF EXISTS close_batch;
+CREATE PROCEDURE close_batch(
+  IN in_batch_id BIGINT,
+  IN in_timestamp BIGINT
+)
+BEGIN
+  DECLARE cur_batch_closed BOOLEAN;
+  DECLARE expected_n_jobs INT;
+  DECLARE staging_n_jobs INT;
+  DECLARE staging_n_ready_jobs INT;
+  DECLARE staging_ready_cores_mcpu INT;
+  DECLARE cur_user VARCHAR(100);
+
+  START TRANSACTION;
+
+  SELECT n_jobs, closed INTO expected_n_jobs, cur_batch_closed FROM batches
+  WHERE id = in_batch_id AND NOT deleted;
+
+  IF cur_batch_closed = 1 THEN
+    COMMIT;
+    SELECT 0 as rc;
+  ELSEIF cur_batch_closed = 0 THEN
```

Comment:
I think the final error is overkill, I think you can just do `IF cur_batch_closed THEN ... ELSE ... END IF` safely.